From 22f998e29d9934402c1811576230842b30af27fb Mon Sep 17 00:00:00 2001
From: Enric Balletbo i Serra <eballetbo@iseebcn.com>
Date: Thu, 8 Sep 2011 11:20:12 +0200
Subject: [PATCH 14/44] OMAP3: igep0020: Fix mux when using board.ei485 kernel parameter

Is too early for use omap_mux_init_signal function in early_param. The
mux_init call just fails with Could not set signal. This patch sets a
flag when board.ei485=no is passed in kernel cmdline, then uses this flag
to set the mux acording to the parameter in board init function.

Signed-off-by: Enric Balletbo i Serra <eballetbo@iseebcn.com>
---
 arch/arm/mach-omap2/board-igep0020.c |   60 +++++++++++++++++++++++-----------
 1 files changed, 41 insertions(+), 19 deletions(-)

diff --git a/arch/arm/mach-omap2/board-igep0020.c b/arch/arm/mach-omap2/board-igep0020.c
index 139fe35..fa4284f 100644
--- a/arch/arm/mach-omap2/board-igep0020.c
+++ b/arch/arm/mach-omap2/board-igep0020.c
@@ -63,6 +63,17 @@
 #define IGEP3_GPIO_LED1_RED	16
 #define IGEP3_GPIO_USBH_NRESET  183
 
+#define IGEP0020_BOARD_OPT_RS232	(1 << 0)
+
+/* Board configuration options */
+struct board_platform_data {
+	unsigned options;
+};
+
+struct board_platform_data igep0020_board_pdata = {
+	.options = 0,
+};
+
 /*
  * IGEP2 Hardware Revision Table
  *
@@ -286,7 +297,6 @@ static const struct usbhs_omap_board_data igep2_usbhs_bdata __initconst = {
 	.reset_gpio_port[2] = -EINVAL,
 };
 
-#ifdef CONFIG_OMAP_MUX
 static struct omap_board_mux board_mux[] __initdata = {
 	/* SMSC9221 LAN Controller */
 	OMAP3_MUX(MCSPI1_CS2, OMAP_MUX_MODE4 | OMAP_PIN_INPUT),
@@ -326,11 +336,25 @@ static struct omap_board_mux board_mux[] __initdata = {
 	OMAP3_MUX(MCBSP2_DR, OMAP_MUX_MODE0 | OMAP_PIN_INPUT),
 	OMAP3_MUX(MCBSP2_DX, OMAP_MUX_MODE0 | OMAP_PIN_OUTPUT),
 	/* Serial ports */
+	OMAP3_MUX(UART1_TX, OMAP_MUX_MODE0 | OMAP_PIN_OUTPUT),
+	OMAP3_MUX(UART1_RX, OMAP_MUX_MODE0 | OMAP_PIN_INPUT_PULLDOWN),
+	OMAP3_MUX(UART1_RTS, OMAP_MUX_MODE0 | OMAP_PIN_OUTPUT),
+	OMAP3_MUX(UART1_CTS, OMAP_MUX_MODE0 | OMAP_PIN_INPUT_PULLDOWN),
 	OMAP3_MUX(UART2_TX, OMAP_MUX_MODE0 | OMAP_PIN_OUTPUT),
 	OMAP3_MUX(UART2_RX, OMAP_MUX_MODE0 | OMAP_PIN_INPUT),
 	{ .reg_offset = OMAP_MUX_TERMINATOR },
 };
-#endif
+
+/* Use UART1 as RS232 (not RS485) */
+static struct omap_board_mux uart1_as_rs232_mux[] __initdata = {
+	OMAP3_MUX(UART1_TX, OMAP_MUX_MODE0 | OMAP_PIN_OUTPUT),
+	OMAP3_MUX(UART1_RX, OMAP_MUX_MODE0 | OMAP_PIN_INPUT_PULLDOWN),
+	OMAP3_MUX(UART1_RTS, OMAP_MUX_MODE7 | OMAP_PIN_INPUT),
+	OMAP3_MUX(UART1_CTS, OMAP_MUX_MODE7 | OMAP_PIN_INPUT),
+	{ .reg_offset = OMAP_MUX_TERMINATOR },
+};
+
+static struct omap_mux_partition *partition;
 
 static int __init ei485_early_param(char *str)
 {
@@ -341,22 +365,12 @@ static int __init ei485_early_param(char *str)
 
 	strncpy(opt, str, 16);
 
-	omap_mux_init_signal("uart1_tx", OMAP_PIN_OUTPUT);
-	omap_mux_init_signal("uart1_rx", OMAP_PIN_INPUT);
-
-	if (!strcmp(opt, "no")) {
-		/*
-		 * To use UART1 as RS232 port instead of RS485 we need configure
-		 * UART1_RTS and UART1_CTS pins in safe mode. To set the RS485
-		 * comment next to lines.
-		 */
-		omap_mux_init_signal("uart1_rts.safe_mode", 0);
-		omap_mux_init_signal("uart1_cts.safe_mode", 0);
-		pr_info("IGEP: board options: ei485=no \n");
-	} else {
-		omap_mux_init_signal("uart1_rts", 0);
-		omap_mux_init_signal("uart1_cts", 0);
-	}
+	/*
+	 * To use UART1 as RS232 port instead of RS485 we need configure
+	 * UART1_RTS and UART1_CTS pins in safe mode.
+	 */
+	if (!strcmp(opt, "no"))
+		igep0020_board_pdata.options  |= IGEP0020_BOARD_OPT_RS232;
 
 	return 0;
 }
@@ -368,7 +382,7 @@ extern void __init igep0022_init(void);
 
 static void __init igep_init(void)
 {
-	int opt;
+	int opt = 0;
 
 	omap3_mux_init(board_mux, OMAP_PACKAGE_CBB);
 	/* Ensure msecure is mux'd to be able to set the RTC. */
@@ -417,6 +431,14 @@ static void __init igep_init(void)
 
 	igep0020_leds_init();
 
+	/*
+	 * By default UART1 is configured for RS485 usage, configure for
+	 * RS232 if flag is enabled.
+	 */
+	partition = omap_mux_get("core");
+	if (igep0020_board_pdata.options & IGEP0020_BOARD_OPT_RS232)
+		omap_mux_write_array(partition, uart1_as_rs232_mux);
+
 	/* Expansion board initialitzations */
 	/* - IGEP0022 */
 	if (igep00x0_buddy_pdata.model == IGEP00X0_BUDDY_IGEP0022)
-- 
1.7.4.1

